### DISCUSSION TITLE
Extracting the SDL without running the app

### DISCUSSION THREAD

**Discussion by badsyntax** [ChilliCream/hotchocolate#4862] [Q&A] [answered]
Hi. I need the SDL to generate client-side code. I'm aware of the `https://localhost:7161/graphql?sdl` endpoint but I don't want to have to run the app to extract the SDL, as I want to generate client side code in CI/CD.

When using a REST/Swagger API, i can extract the OpenAPI spec with: `dotnet swagger tofile --output api-spec.json ./bin/Release/net6.0/MyApp.WebAPI.dll v1.0`

Is there something similar to extract the SDL for GraphQL?

**Comment by tobias-tengler** [comment, marked as answer]
You will not get around running your executable in some way. What you could do, is add a CLI argument to your application and if it's set just write the schema to a file and exit the application.

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services
    .AddGraphQLServer();
   // ...

var app = builder.Build();

if (args.Any(a => a == "--export"))
{
    var executor = await app.Services
         .GetRequiredService<IRequestExecutorResolver>()
         .GetRequestExecutorAsync();

    string schema = executor.Schema.Print();

    // write it to a file, etc.
    return;
}

app.MapGraphQL();
// ...
```

Then you just need to run your application with the argument:
```bash
dotnet run -- --export
```

**Reply by badsyntax** [reply]
Thanks @tobias-tengler I might have issues running the app in the pipelines as i don't think the pipelines runner user has access to various Azure services we require for the app to start. I will generate my client-side code in a different way though. Thanks!

**Reply by ernestfolch** [reply]
Hi @badsyntax, how finally you managed it? (I need to generate the SDL file in a Azure pipeline)

**Reply by michaelstaib** [reply]
Actually you should have no problems with running that. 

The schema creation has no dependency on your services and just relies on internal Hot Chocolate services. So this should not be an issue.

**Reply by badsyntax** [reply]
In my case I changed my approach entirely and I no longer have the need to extract the SDL (but as @michaelstaib clarified there perhaps won't be any probs in running the app to extract the SDL).

Here are my reasons for not requiring SDL in my pipelines: Before using GraphQL I was generating a TypeScript HTTP client in my pipelines and publishing the package to a private npm registry, which made sense as the REST API had a strict schema. In the case of GraphQL, the "schema" is dependant on what the client is asking for, so it doesn't make sense to attempt to publish a generic typed GraphQL client. Instead I'm now generating my GraphQL HTTP client within my front-end app. The client generation uses [graphql-code-generator](https://www.graphql-code-generator.com/) with the [typescript-react-query plugin](https://www.graphql-code-generator.com/plugins/typescript/typescript-react-query) which generates typed react hooks based on client queries. It uses a deployed instance of hotchocolate to extract the graphql types. This process works well for us.

**Reply by ernestfolch** [reply]
Many thanks @badsyntax for your detailed explanation, I got the point. In my case I need the SDL file to import it to our Azure Api Management in our release pipeline. I think I will go ahead with the solution proposed here because, as @michaelstaib clarified, the schema creation has no dependency on our app services, so it won't be any problem in running the app in our pipeline to extract the SDL and generate the artifacts needed in our release pipeline.
